src/events/broker.py

```diff
@@ -0,0 +1,328 @@
+
+        self.balancer_manager = BalancerManager(self)
+
+        self.framework.observe(getattr(self.charm.on, "install"), self._on_install)
+        self.framework.observe(getattr(self.charm.on, "start"), self._on_start)
+        self.framework.observe(getattr(self.charm.on, "config_changed"), self._on_config_changed)
+        self.framework.observe(getattr(self.charm.on, "update_status"), self._on_update_status)
+        self.framework.observe(getattr(self.charm.on, "secret_changed"), self._on_secret_changed)
+
+        self.framework.observe(self.charm.on[PEER].relation_changed, self._on_config_changed)
+
+        self.framework.observe(
+            getattr(self.charm.on, "data_storage_attached"), self._on_storage_attached
+        )
+        self.framework.observe(
+            getattr(self.charm.on, "data_storage_detaching"), self._on_storage_detaching
+        )
+
+    def _on_install(self, _: InstallEvent) -> None:
+        """Handler for `install` event."""
+        self.config_manager.set_environment()
+        self.charm.unit.set_workload_version(self.workload.get_version())
+
+    def _on_start(self, event: StartEvent) -> None:
+        """Handler for `start` event."""
+        if self.charm.state.peer_relation:
+            self.charm.state.unit_broker.update({"cores": str(self.balancer_manager.cores)})
+
+        self.charm._set_status(self.charm.state.ready_to_start)
+        if not isinstance(self.charm.unit.status, ActiveStatus):
+            event.defer()
+            return
+
+        # required settings given zookeeper connection config has been created
+        self.config_manager.set_zk_jaas_config()
+        self.config_manager.set_server_properties()
+        self.config_manager.set_client_properties()
+
+        # start kafka service
+        self.workload.start()
+        logger.info("Kafka snap started")
+
+        # check for connection
+        self.charm.on.update_status.emit()
+
+        # only log once on successful 'on-start' run
+        if isinstance(self.charm.unit.status, ActiveStatus):
+            logger.info(f'Broker {self.charm.unit.name.split("/")[1]} connected')
+
+    def _on_config_changed(self, event: EventBase) -> None:
+        """Generic handler for most `config_changed` events across relations."""
+        if self.charm.state.balancer and self.charm.unit.is_leader():
+            self.charm.state.balancer.update(
+                {
+                    "broker-capacities": self.config_manager.broker_capacities,
+                    "rack_aware": json.dumps(bool(self.config_manager.rack_properties)),
+                }
+            )
+
+        if (
+            not self.upgrade.idle
+            and not self.charm.state.cluster.internal_user_credentials
+            and self.charm.state.zookeeper.zookeeper_connected
+        ):
+            # Create credentials for missing internal user, to reconcile state during upgrades
+            self.zookeeper._on_zookeeper_changed(event)  # type: ignore
```

Comment:
Yes, it's a new addition that I added to make the upgrade integration test pass. The gist of it is: since we now have a new `balancer` internal user, we need to find a way to create it when we update the charm from the latest `stable` or `edge`. Otherwise, the application would get stuck on NO_BROKER_CREDS.

I don't really like it, though. I would prefer any kind of reconciliation logic to be in `upgrade.py`

Reply:
I think we should put this in the `events/upgrade.py` section, something like:
```python
def apply_backwards_compatibility_fixes(self) -> None:
        """A range of functions needed for backwards compatibility."""
        # Rev.113 - needed for allowing log-level upgrades (DPE-123)
        self.charm.config_manager.set_server_jvmflags()
```